### ISSUE TITLE
Collections.files mehods - change in behavior between 4.X and 5.X version

### ISSUE THREAD

**Issue by Jonatha1983** [gradle/gradle#8131]
When using one of the 'files' methods of Configuration class it act differently since gradle 5

### Expected Behavior
When using one of the 'files' method in the Configuration calss:

```
Set<File> files(Closure dependencySpecClosure);
Set<File> files(Spec<? super Dependency> dependencySpec);
Set<File> files(Dependency... dependencies);
```

and the parameter passed to the method is created with the transitive property set to 'False'  the method should not return the transitive dependency as he did not in versions prior 5.X

### Current Behavior
Getting transitive dependencies

### Context
We getting Run time exception due to extra jars in our release folder.

### Steps to Reproduce (for bugs)
Create a project with the content below:

```
plugins` {
    id 'java'
}

repositories { jcenter() }

configurations { myConfig }


Dependency[] deps1 = project.getDependencies().add(configurations.myConfig.name,
                                       [group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.8.9']) {
    transitive = false
}

Dependency[] deps2 = project.getDependencies().add(configurations.myConfig.name, [group: 'io.vertx', name: 'vertx-core', version: '3.5.0']) 

tasks.register("copyJackson", Copy) {
    from configurations.myConfig.files(deps1)
    into project.buildDir.toPath().resolve(gradle.gradleVersion).resolve("3rd").resolve("jackson")
}

tasks.register("copyVertex", Copy) {
    from configurations.myConfig.files(deps2)
    into project.buildDir.toPath().resolve(gradle.gradleVersion).resolve("3rd").resolve("vertex")
}


tasks.register("myCopy"){
    dependsOn tasks.named("copyJackson")
    dependsOn tasks.named("copyVertex")
}
```

Run myCopy task once with gradle version 4.X and with gradle version 5.X and compare the two outputs folder.  
### Your Environment
------------------------------------------------------------
Gradle 5.1-rc-3
------------------------------------------------------------

Build time:   2018-12-21 23:03:45 UTC
Revision:     bf267654ac894c160b5b6c5e5c06d72f1ecaa23c

Kotlin DSL:   1.1.0
Kotlin:       1.3.11
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018
JVM:          1.8.0_101 (Oracle Corporation 25.101-b13)
OS:           Windows 7 6.1 amd64

The Scan result for the run with 4.9
[https://gradle.com/s/nu6c645tmyvgk](url)

The scan result for run with 5.1.3
[https://gradle.com/s/coffnvb4kkyas](url)

**Comment by bigdaz** [comment]
Thanks for the detailed report. I've investigated this change in behaviour, which was triggered when we made `IMPROVED_POM_SUPPORT` the default in Gradle 5.0.

The new behaviour is what we'd expect: you are asking for the files related to the `jackson-databind` module, and this dependency is considered `transitive` since one of the edges pointing at `jackson-databind` is marked as transitive (the edge from `myConfig -> vertx-core -> jackson-databind`).

Prior to this change, the direct dependency on `jackson-databind` resolved to the "default" configuration/variant for the component. The transitive dependency brought in via `vertx-core` resolved to the "runtime" configuration/variant. This meant that Gradle did not merge the resolved nodes correctly, providing the behaviour you see in Gradle < 5.0.

Some breaking changes were expected with the enabling of `IMPROVED_POM_SUPPORT`, and this is an example of such a change. If you can describe your use case in a bit more detail, I'm sure we can help come up with a solution that works with Gradle 5.0.

**Comment by Jonatha1983** [comment]
Thanks @bigdaz   

We support some legacy code.

We make use of Gradle and its dependency coordinates syntax to define, download  and then copy dependencies to some expected folders.
In one project we define all the dependencies and we create dynamically tasks to copy each dependencies into a dedicate folder. 
The set of dependencies must be added to a single configuration for conflicts resolutions.
and so we create one and add those deps to it. 
We then create a task for each dependency (or set of dependencies) when we define its inputs files to be the configurations files (after filtering with its deps) and as outputs dir the dedicate folder to copy to.

So some of those dependencies were set to transitive False and while in gradle version <5 we found the transitive working as we expected in >=5 we saw some extra jars that eventually cause us some class path problems and RunTimeExceptiosns.

Any suggestion on how we can achieve the same result in which dependencies that were define as not transitive will remain such  ?

Thanks ahead 
Jonathan

**Comment by Jonatha1983** [comment]
Hi @bigdaz  just a small clarification:  

I tried to find a solution at: 
[https://docs.gradle.org/current/userguide/managing_transitive_dependencies.html#sec:excluding_transitive_module_dependencies
](url)

But also the class documentation says:

> Excluding a particular transitive dependency does not guarantee that it does not show up
     * in the dependencies of a given configuration.
     * For example, some other dependency, which does not have any exclude rules,
     * might pull in exactly the same transitive dependency.


So to summarize: we cannot exclude this dep for the entire configuration and we must exclude for a specific dependency in that configuration - what is the best available solution ?  thanks ahead

**Comment by bigdaz** [comment]
OK thanks for the clarification. I tried a few more things and the change in behaviour in 5.x is quite odd.

I changed the declaration of `deps1` to target the "runtime" `configuration` of jackson ie:
```
Dependency[] deps1 = project.getDependencies().add(configurations.myConfig.name,
                                       [group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.8.9', configuration: 'runtime']) {
    transitive = false
}
```

When I do this, then the behaviour in 4.x and 5.x is _reversed_: the transitive dependencies are _included_ when resolving with 4.x and _excluded_ when resolving with 5.x.

I'm not sure if that helps you much: if you consistently add `configuration: 'runtime'` to all of your dependency declarations you should get the same results with Gradle 5.x as you see currently with 4.x.

Thanks for reporting this issue. We'll add it to the backlog of things to fix.
